src/devices/sound/st0032.cpp

```diff
@@ -0,0 +1,240 @@
+// license:BSD-3-Clause
+// copyright-holders:Tomasz Slanina
+/************************************
+      Seta custom ST-0032 chip (sound emulation)
+      based on ST-0026 Nile emulation by Tomasz Slanina
+      similar as Nile sound, but it's has 16 voices
+
+      Register Format (32 byte per voices)
+
+      00-1f: Voice 0
+      Offset Bit               Description
+             fedcba98 76543210 
+      04     xxxxxxxx xxxxxxxx Start position LSB
+      06     xxxxxxxx xxxxxxxx Start position MSB
+      0a     -------- -----xx- Used but unknown
+             -------- -------x Loop enable
+      0c     xxxxxxxx xxxxxxxx Frequency
+      0e     xxxxxxxx xxxxxxxx Loop Start position LSB
+      12     xxxxxxxx xxxxxxxx Loop Start position MSB
+      14     xxxxxxxx xxxxxxxx Loop End position LSB
+      16     xxxxxxxx xxxxxxxx Loop End position MSB
+      18     xxxxxxxx xxxxxxxx End position LSB
+      1a     xxxxxxxx xxxxxxxx End position MSB
+      1c     xxxxxxxx xxxxxxxx Right Volume
+      1e     xxxxxxxx xxxxxxxx Left Volume
+
+      20-3f: Voice 1
+      ...
+      1e0-1ff: Voice 15
+
+      200: Keyon/off, Bit 0-15 means Voice 0-15
+      210: Used but unknown
+
+      Other registers are unknown/unused
+
+      TODO:
+      - Verify loop behavior from real hardware
+
+************************************/
+
+#include "emu.h"
+#include "st0032.h"
+
+enum
+{
+	ST0032_REG_UNK0=0,
+	ST0032_REG_SPTR_TOP,
+	ST0032_REG_SPTR_LO,
+	ST0032_REG_SPTR_HI,
+	ST0032_REG_UNK_4,
+	ST0032_REG_FLAGS,
+	ST0032_REG_FREQ,
+	ST0032_REG_LSPTR_LO,
+	ST0032_REG_UNK_8,
+	ST0032_REG_LSPTR_HI,
+	ST0032_REG_LEPTR_LO,
+	ST0032_REG_LEPTR_HI,
+	ST0032_REG_EPTR_LO,
+	ST0032_REG_EPTR_HI,
+	ST0032_REG_VOL_R,
+	ST0032_REG_VOL_L
+};
+
+
+DEFINE_DEVICE_TYPE(ST0032_SOUND, st0032_sound_device, "st0032_sound", "Seta ST-0032 Sound Hardware")
+
+st0032_sound_device::st0032_sound_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, ST0032_SOUND, tag, owner, clock),
+		device_sound_interface(mconfig, *this),
+		m_stream(nullptr),
+		m_sound_ram(*this, DEVICE_SELF),
```

Comment:
Is this actually RAM, or is it ROM?  If it’s RAM you shouldn’t be using a region for it like this.  If it’s ROM, you shouldn’t really be working with non-const pointers to it in `sound_stream_update`.

Reply:
ST0032 sound samples are fetched from ROM, that is just copy-paste error from nile.cpp(also uses ROM for sound samples).